### DISCUSSION TITLE
[Docs] Using context in Client Components

### DISCUSSION THREAD

**Discussion by doksara** [vercel/next.js#52751] [Help]
### Summary

So, I have moved to next 13.4 and trying to wrap my head around the new Server / Client components separation and rendering. In the section [The use client directive](https://nextjs.org/docs/getting-started/react-essentials#the-use-client-directive) it says:

_"Once "use client" is defined in a file, all other modules imported into it, including child components, are considered part of the client bundle."_

If I understood this correctly, it means if we use the '`use client'` directive in our root layout, literally all of our code will be shipped in the client bundle? If this is correct, how do we make sense of React context and context providers then? Most of the providers (i.e. `ThemeProvider`) are defined at the root of our component tree which means we are forced to opt-in to Client component and then the whole idea of [server-first](https://nextjs.org/docs/getting-started/react-essentials#thinking-in-server-components) does not make sense anymore:

![thinking-in-server-components (1)](https://github.com/vercel/next.js/assets/32953370/2b543c5d-9592-4c2a-bbcc-1526c087161a)

**Comment by icyJoseph** [comment]
The Server Components and "use client" components can be weaved/patched/joined by React, so that the resulting HTML is congruent with what your JSX description wanted.

That is to say that if you, at the root layout, import a Providers component, which is itself a use client component, you create a Client Side Tree that'll be weaved into the whatever the root layout itself renders, and whatever potential server components or client components down the tree render as well.

So any client components down in the tree can still use context. However, server components, because they lack interactivity won't have access to such data layer. 

- https://nextjs.org/docs/getting-started/react-essentials#using-context-in-client-components

Notice how, in the link above, the code snippet for the layout itself, does not use `use client` directive.
